pkg/mcs/resourcemanager/server/keyspace_manager.go

```diff
@@ -376,6 +375,64 @@ func (rt *ruTracker) getRUPerSec() float64 {
 	return rt.lastEMA
 }
 
+// groupRUTracker is used to track the RU consumption of a resource group.
+// It matains a map of RU trackers for each client unique ID.
+type groupRUTracker struct {
+	syncutil.RWMutex
+	ruTrackers map[uint64]*ruTracker
+}
+
+func newGroupRUTracker() *groupRUTracker {
+	return &groupRUTracker{
+		ruTrackers: make(map[uint64]*ruTracker),
+	}
+}
+
+func (grt *groupRUTracker) getOrCreateRUTracker(clientUniqueID uint64) *ruTracker {
+	grt.RLock()
```

Comment:
for this case, I think the singleflight https://pkg.go.dev/golang.org/x/sync/singleflight is very suitable.